**PKA**

**NORMAL**

In [38]:
PKA_FEATURES = ['RPCS', 'angle_R1X1R2', 'angle_X2X1R1', 'angle_X1X2R2', 'angle_R2X2R1', 
                'distance_between_atoms_in_cycle_and_f_group', 
                'distance_between_atoms_in_f_group_centers', 'PBF', 'mol_weight', 'dipole_moment', 'PPSA5',
                'avg_atoms_in_cycle', 'nHRing', 'cis/trans', 'FPSA3', 'nF', 'chirality',
                'sasa', 'PNSA5', 'GeomShapeIndex', 'TASA', 'mol_num_cycles',
                'f_freedom', 'nFRing', 'identificator', 'nO', 'nARing', 'nC', 'nFHRing',
                'f_to_fg']

In [39]:
import pandas as pd

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\updated_features\csv_for_rulefit\train_pKa_v5_features_canon_smiles.csv', index_col=0)

y = train['pKa']
X=train[PKA_FEATURES]

test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\updated_features\csv_for_rulefit\test_pKa_v5_features_canon_smiles.csv', index_col=0)

In [40]:
cv_indices_dict = {0: [], 1: []}
index = 0
for _, row in train.iterrows():
    cv_indices_dict[row['fold_id']].append(index)
    index += 1
cv_indices = [[cv_indices_dict[0], cv_indices_dict[1]], [cv_indices_dict[1], cv_indices_dict[0]]]

In [42]:
from sklearn.linear_model import ElasticNetCV
from sklearn.metrics import r2_score
import pandas as pd
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_error as mse

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\updated_features\csv_for_rulefit\train_pKa_v5_features_canon_smiles.csv', index_col=0)
y = train['pKa']
X=train.drop(['pKa', 'fold_id'], axis=1)

regr = ElasticNetCV(cv=cv_indices, random_state=0, max_iter=20000, n_alphas=1000)
regr.fit(X, y)
test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\updated_features\csv_for_rulefit\test_pKa_v5_features_canon_smiles.csv', index_col=0)
def metrics(test, feature):
    y_pred = regr.predict(test.drop([feature], axis=1))
    r2 = r2_score(y_pred=y_pred, y_true = test[feature])
    print(f'=========%s========='%(feature))
    print('R^2 = '+str(round(r2, 3)))
    print('MAE = ', round(mae(y_true = test[feature], y_pred=y_pred), 3))
    print('MSE = ', round(mse(y_true = test[feature], y_pred=y_pred), 3))
metrics(test, 'pKa')
regr.score(X, y)
metrics(train.drop('fold_id', axis=1), 'pKa')

=========pKa=========
R^2 = 0.878
MAE =  0.691
MSE =  0.7
=========pKa=========
R^2 = 0.896
MAE =  0.68
MSE =  0.686


In [43]:
from sklearn.datasets import make_regression
from sklearn.linear_model import ElasticNetCV
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import r2_score, mean_absolute_error

from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from math import sqrt

def calculate_metrics(true_values, pred_values):
    mse = round(mean_squared_error(true_values, pred_values),3)
    mae = round(mean_absolute_error(true_values, pred_values),3)
    r_score = round(r2_score(true_values, pred_values),3)

    return {"mse": mse,
            "mae": mae,
            "r^2": r_score,}

# Initialize ElasticNetCV
model = ElasticNetCV(cv=2, random_state=0, max_iter=20000, n_alphas=1000)

# Fit the model
model.fit(X, y)

# Get cross-validated predictions
y_pred_cv = cross_val_predict(model, X, y, cv=cv_indices)

print(calculate_metrics(true_values=y, pred_values=y_pred_cv))

def metrics(test, feature):
    y_pred = model.predict(test.drop([feature], axis=1))
    r2 = r2_score(y_pred=y_pred, y_true = test[feature])
    print(f'=========%s========='%(feature))
    print('R^2 = '+str(round(r2, 3)))
    print('MAE = ', round(mae(y_true = test[feature], y_pred=y_pred), 3))
    print('MSE = ', round(mse(y_true = test[feature], y_pred=y_pred), 3))
metrics(test, 'pKa')
model.score(X, y)
print("Train")
metrics(train.drop('fold_id', axis=1), 'pKa')

{'mse': 0.868, 'mae': 0.757, 'r^2': 0.868}
=========pKa=========
R^2 = 0.878
MAE =  0.691
MSE =  0.7
Train
=========pKa=========
R^2 = 0.896
MAE =  0.68
MSE =  0.686


**MINMAX scaler**

In [25]:
PKA_FEATURES = ['RPCS', 'angle_R1X1R2', 'angle_X2X1R1', 'angle_X1X2R2', 'angle_R2X2R1', 
                'distance_between_atoms_in_cycle_and_f_group', 
                'distance_between_atoms_in_f_group_centers', 'PBF', 'mol_weight', 'dipole_moment', 'PPSA5',
                'avg_atoms_in_cycle', 'nHRing', 'cis/trans', 'FPSA3', 'nF', 'chirality',
                'sasa', 'PNSA5', 'GeomShapeIndex', 'TASA', 'mol_num_cycles',
                'f_freedom', 'nFRing', 'identificator', 'nO', 'nARing', 'nC', 'nFHRing',
                'f_to_fg']

In [26]:
import pandas as pd

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\train_pka_minmax_only_non_categorizal_features_scaled.csv', index_col=0)

y = train['pKa']
X=train[PKA_FEATURES]

test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\test_pka_minmax_only_non_categorizal_features_scaled.csv', index_col=0)

In [27]:
cv_indices_dict = {0: [], 1: []}
index = 0
for _, row in train.iterrows():
    cv_indices_dict[row['fold_id']].append(index)
    index += 1
cv_indices = [[cv_indices_dict[0], cv_indices_dict[1]], [cv_indices_dict[1], cv_indices_dict[0]]]

In [28]:
from sklearn.linear_model import ElasticNetCV
from sklearn.metrics import r2_score
import pandas as pd
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import mean_squared_error as mse

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\train_pka_minmax_only_non_categorizal_features_scaled.csv', index_col=0)
y = train['pKa']
X=train.drop(['pKa', 'fold_id'], axis=1)

regr = ElasticNetCV(cv=cv_indices, random_state=0, max_iter=20000, n_alphas=1000)
regr.fit(X, y)
test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\test_pka_minmax_only_non_categorizal_features_scaled.csv', index_col=0)

def metrics(test, feature):
    y_pred = regr.predict(test.drop([feature], axis=1))
    r2 = r2_score(y_pred=y_pred, y_true = test[feature])
    print(f'=========%s========='%(feature))
    print('R^2 = '+str(round(r2, 3)))
    print('MAE = ', round(mae(y_true = test[feature], y_pred=y_pred), 3))
    print('MSE = ', round(mse(y_true = test[feature], y_pred=y_pred), 3))
metrics(test, 'pKa')
regr.score(X, y)
metrics(train.drop('fold_id', axis=1), 'pKa')

=========pKa=========
R^2 = 0.903
MAE =  0.619
MSE =  0.558
=========pKa=========
R^2 = 0.939
MAE =  0.516
MSE =  0.404


In [29]:
from sklearn.datasets import make_regression
from sklearn.linear_model import ElasticNetCV
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import r2_score, mean_absolute_error

from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from math import sqrt

def calculate_metrics(true_values, pred_values):
    mse = round(mean_squared_error(true_values, pred_values),3)
    mae = round(mean_absolute_error(true_values, pred_values),3)
    r_score = round(r2_score(true_values, pred_values),3)

    return {"mse": mse,
            "mae": mae,
            "r^2": r_score,}

# Initialize ElasticNetCV
model = ElasticNetCV(cv=2, random_state=0, max_iter=20000, n_alphas=1000)

# Fit the model
model.fit(X, y)

# Get cross-validated predictions
y_pred_cv = cross_val_predict(model, X, y, cv=cv_indices)

print(calculate_metrics(true_values=y, pred_values=y_pred_cv))

def metrics(test, feature):
    y_pred = model.predict(test.drop([feature], axis=1))
    r2 = r2_score(y_pred=y_pred, y_true = test[feature])
    print(f'=========%s========='%(feature))
    print('R^2 = '+str(round(r2, 3)))
    print('MAE = ', round(mae(y_true = test[feature], y_pred=y_pred), 3))
    print('MSE = ', round(mse(y_true = test[feature], y_pred=y_pred), 3))
metrics(test, 'pKa')
model.score(X, y)
print("Train")
metrics(train.drop('fold_id', axis=1), 'pKa')

{'mse': 1.472, 'mae': 1.099, 'r^2': 0.777}
=========pKa=========
R^2 = 0.85
MAE =  0.801
MSE =  0.86
Train
=========pKa=========
R^2 = 0.852
MAE =  0.847
MSE =  0.978


**MINMAX scaler**

In [30]:
import pandas as pd

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\train_pka_standart_scaler_non_categorizal_features_scaled.csv', index_col=0)

y = train['pKa']
X=train[PKA_FEATURES]

test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\test_pka_standart_scaler_non_categorizal_features_scaled.csv', index_col=0)

In [31]:
cv_indices_dict = {0: [], 1: []}
index = 0
for _, row in train.iterrows():
    cv_indices_dict[row['fold_id']].append(index)
    index += 1
cv_indices = [[cv_indices_dict[0], cv_indices_dict[1]], [cv_indices_dict[1], cv_indices_dict[0]]]

In [32]:
from sklearn.datasets import make_regression
from sklearn.linear_model import ElasticNetCV
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import r2_score, mean_absolute_error

from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from math import sqrt

def calculate_metrics(true_values, pred_values):
    mse = round(mean_squared_error(true_values, pred_values),3)
    mae = round(mean_absolute_error(true_values, pred_values),3)
    r_score = round(r2_score(true_values, pred_values),3)

    return {"mse": mse,
            "mae": mae,
            "r^2": r_score,}

train = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\train_pka_standart_scaler_non_categorizal_features_scaled.csv', index_col=0)
y = train['pKa']
X=train.drop(['pKa', 'fold_id'], axis=1)

regr = ElasticNetCV(cv=cv_indices, random_state=0, max_iter=20000, n_alphas=1000)
regr.fit(X, y)
test = pd.read_csv(r'C:\work\DrugDiscovery\main_git\XAI_Chem\data\pKA_logP_minmaxscaler_canon_smiles\pKa\test_pka_standart_scaler_non_categorizal_features_scaled.csv', index_col=0)

# Get cross-validated predictions
y_pred_cv = cross_val_predict(model, X, y, cv=cv_indices)

print(calculate_metrics(true_values=y, pred_values=y_pred_cv))

def metrics(test, feature):
    y_pred = model.predict(test.drop([feature], axis=1))
    r2 = r2_score(y_pred=y_pred, y_true = test[feature])
    print(f'=========%s========='%(feature))
    print('R^2 = '+str(round(r2, 3)))
    print('MAE = ', round(mean_absolute_error(y_true = test[feature], y_pred=y_pred), 3))
    print('MSE = ', round(mean_squared_error(y_true = test[feature], y_pred=y_pred), 3))

metrics(test, 'pKa')
model.score(X, y)

print("Train")
metrics(train.drop('fold_id', axis=1), 'pKa')

{'mse': 0.674, 'mae': 0.66, 'r^2': 0.898}
=========pKa=========
R^2 = 0.85
MAE =  0.8
MSE =  0.86
Train
=========pKa=========
R^2 = 0.852
MAE =  0.847
MSE =  0.978


In [33]:
from sklearn.datasets import make_regression
from sklearn.linear_model import ElasticNetCV
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import r2_score, mean_absolute_error

from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from math import sqrt

def calculate_metrics(true_values, pred_values):
    mse = round(mean_squared_error(true_values, pred_values),3)
    mae = round(mean_absolute_error(true_values, pred_values),3)
    r_score = round(r2_score(true_values, pred_values),3)

    return {"mse": mse,
            "mae": mae,
            "r^2": r_score,}

# Initialize ElasticNetCV
model = ElasticNetCV(cv=2, random_state=0, max_iter=20000, n_alphas=1000)

print("CV:")
amount_of_cv = 2
for cv_index in range(amount_of_cv):
    X_train_cv = X.iloc[cv_indices[cv_index][0]]
    y_train_cv = y.iloc[cv_indices[cv_index][0]]

    model.fit(X_train_cv, y_train_cv)
    y_pred_train = model.predict(X_train_cv)

    X_test_cv = X.iloc[cv_indices[cv_index][1]]
    y_test_cv = y.iloc[cv_indices[cv_index][1]]

    y_pred = model.predict(X_test_cv)

    print("Train:", calculate_metrics(true_values=y_train_cv, pred_values=y_pred_train))
    print("OOS:", calculate_metrics(true_values=y_test_cv, pred_values=y_pred))

CV:
Train: {'mse': 0.351, 'mae': 0.481, 'r^2': 0.948}
OOS: {'mse': 0.462, 'mae': 0.561, 'r^2': 0.928}
Train: {'mse': 0.457, 'mae': 0.56, 'r^2': 0.928}
OOS: {'mse': 0.889, 'mae': 0.759, 'r^2': 0.869}


In [34]:
model

ElasticNetCV(cv=2, max_iter=20000, n_alphas=1000, random_state=0)

NORMAL:
{'mse': 0.868, 'mae': 0.757, 'r^2': 0.868}
=========pKa=========
R^2 = 0.878
MAE =  0.691
MSE =  0.7
Train
=========pKa=========
R^2 = 0.896
MAE =  0.68
MSE =  0.686



MINMAX:
{'mse': 1.472, 'mae': 1.099, 'r^2': 0.777}
=========pKa=========
R^2 = 0.85
MAE =  0.801
MSE =  0.86
Train
=========pKa=========
R^2 = 0.852
MAE =  0.847
MSE =  0.978




STANDART SCALER:
{'mse': 0.674, 'mae': 0.66, 'r^2': 0.898}
=========pKa=========
R^2 = 0.85
MAE =  0.8
MSE =  0.86
Train
=========pKa=========
R^2 = 0.852
MAE =  0.847
MSE =  0.978